File: solutions/python/locomotive-engineer/1/locomotive_engineer.py
def get_list_of_wagons(*args):
    return list(args)

def fix_list_of_wagons(wagons, missing_wagons):
    # Unpack first two wagons and the rest
    first, second, *rest = wagons
    
    # Find the locomotive (1) in the remaining wagons
    for i, wagon in enumerate(rest):
        if wagon == 1:
            # Reconstruct the list with missing wagons inserted after locomotive
            return [1, *missing_wagons, *rest[i+1:], first, second]
    return wagons  # fallback if locomotive not found
 
def add_missing_stops(routing, **stops):
    # Extract stop numbers and sort them
    stop_numbers = sorted(int(key.split('_')[1]) for key in stops if key.startswith('stop_'))
    # Create ordered list of stops
    routing['stops'] = [stops[f'stop_{num}'] for num in stop_numbers]
    return routing   
    
def extend_route_information(route, more_route_information):
    return {**route, **more_route_information}

def fix_wagon_depot(wagons_rows):
    # Unpack the three color rows
    red_row, blue_row, orange_row = wagons_rows
    
    # Unpack each row's wagons
    (red1, red2, red3) = red_row
    (blue1, blue2, blue3) = blue_row
    (orange1, orange2, orange3) = orange_row
    
    # Reconstruct the grid by color columns
    return [
        [red1, blue1, orange1],
        [red2, blue2, orange2],
        [red3, blue3, orange3]
    ]
def fix_wagon_depot(wagons_rows):
    # Using zip to transpose the rows into columns
    return [list(column) for column in zip(*wagons_rows)]

File: solutions/ruby/locomotive-engineer/1/locomotive_engineer.rb
class LocomotiveEngineer
  def self.generate_list_of_wagons(*wagons)
    wagons
  end

  def self.fix_list_of_wagons(each_wagons_id, missing_wagons)
    # Find position of locomotive (always 1)
    loco_index = each_wagons_id.index(1)

    # The two wagons immediately before the locomotive are the ones to move to end
    to_move = each_wagons_id[loco_index - 2, 2]   # [index-2, index-1]

    # Everything after locomotive stays in place (after inserted missing wagons)
    after_loco = each_wagons_id[loco_index + 1..-1]

    # Rebuild:
    # 1 + missing wagons + wagons after 1 + the two moved wagons
    [1, *missing_wagons, *after_loco, *to_move]
  end

  def self.add_missing_stops(route, **stops_keywords)
    stop_cities = stops_keywords
      .sort_by { |k, _| k.to_s[/\d+/].to_i }
      .map { |_, city| city }

    { **route, stops: stop_cities }
  end

  def self.extend_route_information(route, more_route_information)
    { **route, **more_route_information }
  end
end